src/hazelcore/semantics/TextShape.re

```diff
@@ -9,34 +9,46 @@ type t =
   | ExpandingKeyword(ExpandingKeyword.t)
   | Var(Var.t);
 
+let int_of_intlit_opt = (n: string): option(int) => {
+  let stripped_text = StringUtil.strip_underscore(n);
+  int_of_string_opt(stripped_text);
+};
+
+let float_of_floatlit_opt = (f: string): option(float) => {
+  let stripped_text = StringUtil.strip_underscore(f);
+  float_of_string_opt(stripped_text);
+};
+
 let of_text = (text: string): option(t) => {
+  let stripped_text = StringUtil.strip_underscore(text);
   switch (
-    int_of_string_opt(text),
-    float_of_string_opt(text),
+    int_of_string_opt(stripped_text),
+    float_of_string_opt(stripped_text),
     bool_of_string_opt(text),
     ExpandingKeyword.mk(text),
   ) {
   /* the num_digits double count with leading zeros when the int is evaluated to 0  */
   | (Some(n), _, _, _) when n == 0 =>
-    StringUtil.num_leading_zeros(text) == String.length(text)
+    StringUtil.num_leading_zeros(stripped_text)
```

Comment:
not sure I understand why we have these checks related to the length here... can you clarify in a comment? what is an example of a situation where int_of_string_opt returns Some(0) but this check fails and the function returns None?